src/live_restore/live_restore_fs.c

```diff
@@ -1395,90 +1389,240 @@ __live_restore_setup_lr_fh_directory(WT_SESSION_IMPL *session, WTI_LIVE_RESTORE_
+  const char *name, uint32_t flags, WTI_LIVE_RESTORE_FILE_HANDLE *lr_fh, WT_FS_OPEN_FILE_TYPE type,
+  bool dest_exist, bool source_exist)
+{
+    if (!dest_exist && source_exist) {
+        /* Atomically copy across the file. */
+        WT_RET(__live_restore_fs_atomic_copy_file(session, lr_fs, type, name));
+    }
 
-                WT_RET(lr_fh->source->fh_size(lr_fh->source, wt_session, &source_size));
-                __wt_verbose_debug1(session, WT_VERB_LIVE_RESTORE,
-                  "Creating destination file backed by source file. Copying size (%" PRId64
-                  ") from source file",
-                  source_size);
-                lr_fh->source_size = (size_t)source_size;
+    WT_RET(__live_restore_fs_open_in_destination(lr_fs, session, lr_fh, name, flags, !dest_exist));
+    lr_fh->destination.complete = true;
+    return (0);
+}
 
-                /*
-                 * Set size by truncating. This is a positive length truncate so it actually extends
-                 * the file. We're bypassing the live_restore layer so we don't try to modify the
-                 * extents in hole_list_head.
-                 */
-                WT_RET(lr_fh->destination.fh->fh_truncate(
-                  lr_fh->destination.fh, wt_session, source_size));
+/*
+ * __live_restore_setup_lr_fh_file --
+ *     Setup a live restore file handle for a file. This function does some initial file state
+ *     investigation before calling separate functions depending on the type of file.
+ */
+static int
+__live_restore_setup_lr_fh_file(WT_SESSION_IMPL *session, WTI_LIVE_RESTORE_FS *lr_fs,
+  const char *name, WT_FS_OPEN_FILE_TYPE file_type, uint32_t flags,
+  WTI_LIVE_RESTORE_FILE_HANDLE *lr_fh)
+{
+    /*!!!
+     * All non directory open file calls end up here, which means we need to handle:
+     *  - WT_FS_OPEN_FILE_TYPE_CHECKPOINT
+     *  - WT_FS_OPEN_FILE_TYPE_DATA
+     *  - WT_FS_OPEN_FILE_TYPE_LOG
+     *  - WT_FS_OPEN_FILE_TYPE_REGULAR
+     *
+     * Right now we handle everything but the checkpoint type which appears to be unused. Log and
+     * regular files are treated the same in that they are atomically copied on open. Then for any
+     * subsequent open they will be immediately complete.
+     *
+     * Data type files are the b-trees, they are not copied on open and are expected to go through
+     * the extent import path which will initialize the relevant extent lists.
+     */
+    WT_ASSERT(session, file_type != WT_FS_OPEN_FILE_TYPE_CHECKPOINT);
+
+    /*!!!
+     * We need to handle a number of scenario in this function providing us with a somewhat complex
+     * decision tree. The relevant pieces of state for any file are:
+     *   - Whether the live restore is complete or not.
+     *   - Whether a stop file exists for that file.
+     *   - Whether that file exists in the destination or the source.
+     *   - Various flag combinations such as create and exclusive.
```

Comment:
```suggestion
     *   - Flag combinations such as create and exclusive.
```
These are the only flag combinations I see so we shouldn't imply there are others.

Reply:
99ead325d